node/src/components/linear_chain_sync/operations.rs

```diff
@@ -484,19 +522,170 @@ pub(crate) async fn run_fast_sync_task(
+
+    let trusted_block =
+        *fetch_and_store_block_by_hash(effect_builder, trusted_block_header.hash()).await?;
+
+    // Sync to genesis
+    let mut walkback_block = trusted_block.clone();
+    loop {
+        sync_deploys_and_transfers_and_state(effect_builder, &walkback_block).await?;
+        if walkback_block.height() == 0 {
+            break;
+        } else {
+            walkback_block = *fetch_and_store_block_by_hash(
+                effect_builder,
+                *walkback_block.header().parent_hash(),
+            )
+            .await?;
+        }
+    }
+
+    // Sync forward until we are at the current version.
+    let mut most_recent_block = trusted_block;
+    while most_recent_block.header().protocol_version() < chainspec.protocol_config.version {
+        let maybe_fetched_block_with_metadata = fetch_and_store_block_by_height(
+            effect_builder,
+            most_recent_block.header().height() + 1,
+            &trusted_key_block_info,
+            chainspec.highway_config.finality_threshold_fraction,
+        )
+        .await?;
+        most_recent_block = match maybe_fetched_block_with_metadata {
+            Some(block_with_metadata) => block_with_metadata.block,
+            None => {
+                tokio::time::sleep(SLEEP_DURATION_SO_WE_DONT_SPAM).await;
+                continue;
+            }
+        };
+        sync_deploys_and_transfers_and_state(effect_builder, &most_recent_block).await?;
+        if let Some(key_block_info) =
+            KeyBlockInfo::maybe_from_block_header(most_recent_block.header())
+        {
+            trusted_key_block_info = key_block_info;
+        }
+    }
+
+    Ok((trusted_key_block_info, most_recent_block.take_header()))
+}
+
+/// Runs the fast synchronization task.
+pub(crate) async fn run_fast_sync_task(
+    effect_builder: EffectBuilder<JoinerEvent>,
+    trusted_hash: BlockHash,
+    chainspec: Arc<Chainspec>,
+    node_config: NodeConfig,
+) -> Result<BlockHeader, LinearChainSyncError> {
+    // Fetch the trusted header
+    let trusted_block_header = fetch_and_store_block_header(effect_builder, trusted_hash).await?;
+
+    let era_duration: TimeDiff = std::cmp::max(
+        chainspec.highway_config.min_round_length() * chainspec.core_config.minimum_era_height,
+        chainspec.core_config.era_duration,
     );
-    let mut outstanding_trie_keys = vec![Blake2bHash::from(
-        *most_recent_block_header.state_root_hash(),
-    )];
-    while let Some(trie_key) = outstanding_trie_keys.pop() {
-        let missing_descendant_trie_keys =
-            fetch_trie_and_insert_into_trie_store(effect_builder, trie_key).await?;
-        outstanding_trie_keys.extend(missing_descendant_trie_keys);
+
+    if trusted_block_header.timestamp()
+        + era_duration
+            * chainspec
+                .core_config
+                .unbonding_delay
+                .saturating_sub(chainspec.core_config.auction_delay)
```

Comment:
Maybe we can extract it to a method on `chainspec.core_config`? `bonded_eras_num`? @afck